includes/express-checkout/class-wc-payments-express-checkout-button-helper.php

```diff
@@ -385,6 +385,26 @@ public function should_show_express_checkout_button() {
 			return false;
 		}
 
+		// Non-shipping product and billing is calculated based on shopper billing addres.
+		if (
+			// Except for the Pay for order page.
+			! $this->is_pay_for_order_page() &&
+
+			// If the product doesn't needs shipping.
+			(
+				// on the product page.
+				( $this->is_product() && ! $this->product_needs_shipping( $this->get_product() ) ) ||
+
+				// on the cart or checkout page.
+				( ( $this->is_cart() || $this->is_checkout() ) && ! WC()->cart->needs_shipping() )
+			)
+
+			// ...and billing is calculated based on billing address.
+			&& 'billing' === get_option( 'woocommerce_tax_based_on' )
+		) {
+			return false;
+		}
```

Comment:
@asumaran I wonder if we should move this logic a bit down in this function to keep product-specific logic more clustered.

If you move it below [this check](https://github.com/Automattic/woocommerce-payments/blob/76d2cd5199a6cfa61b4a36f42ade13af0bb2a78a/includes/express-checkout/class-wc-payments-express-checkout-button-helper.php#L416-L418) we also wouldn't need to explicitly check if `! $this->is_pay_for_order_page()`, since it would always return `true` for pay for order, but I'd still keep a comment like [this](https://github.com/Automattic/woocommerce-payments/blob/76d2cd5199a6cfa61b4a36f42ade13af0bb2a78a/includes/express-checkout/class-wc-payments-express-checkout-button-helper.php#L421) to explain why.

Reply:
done [13644f9](https://github.com/Automattic/woocommerce-payments/pull/9089/commits/13644f909ba58799bfa7a7158f89e90749acd147)